src/storage/txn/process.rs

```diff
@@ -872,6 +940,254 @@ fn process_write_impl<S: Snapshot, L: LockMgr>(
+            Command::Commit { ref ctx, .. } => ctx.clone(),
+            _ => unreachable!(),
+        };
+        let multi_res = tag == CommandKind::prewrite;
+        let mut txn = MvccTxn::new(snapshot, 0, false)?;
+        // check conflict
+        let results = match tag {
+            CommandKind::prewrite => {
+                let mut results = Vec::new();
+                let len = commands.len();
+                for i in 0..len {
+                    if ids[i].is_none() {
+                        continue;
+                    }
+                    if let Command::Prewrite {
+                        mutations,
+                        primary,
+                        start_ts,
+                        options,
+                        ..
+                    } = &mut commands[i]
+                    {
+                        if options.for_update_ts != 0 {
+                            panic!("batch command does not accept pessimistic prewrite request");
+                        }
+                        txn.start_ts(*start_ts);
+                        let mut locks = Vec::new();
+                        let mut failed = false;
+                        let mutations = std::mem::replace(mutations, vec![]);
+                        for m in mutations {
+                            match txn.prewrite(m, &primary, &options) {
+                                e @ Err(MvccError::KeyIsLocked { .. }) => {
+                                    locks.push(e.map_err(Error::from).map_err(StorageError::from));
+                                }
+                                Err(e) => {
+                                    results.push((
+                                        ids[i].take().unwrap(),
+                                        Msg::FinishedWithErr {
+                                            cid,
+                                            err: Error::from(e),
+                                            tag,
+                                        },
+                                    ));
+                                    failed = true;
+                                    break;
+                                }
+                                _ => {}
+                            }
+                        }
+                        if !failed && !locks.is_empty() {
+                            results.push((
+                                ids[i].take().unwrap(),
+                                Msg::WriteFinished {
+                                    cid,
+                                    pr: ProcessResult::MultiRes { results: locks },
+                                    result: Ok(()),
+                                    tag,
+                                },
+                            ));
+                        }
+                    } else {
+                        unreachable!();
+                    }
+                }
+                results
```

Comment:
Could all of this move to another function?  The function `process_batch_write_impl ` looks like too big....